src/inc/til/rectangle.h

```diff
@@ -0,0 +1,491 @@
+                //
+                //
+                //         |---------------|
+                //         | other         |
+                //         |---------------|
+
+                // We generate these rectangles by the original and intersect points, but some of them might be empty when the intersect
+                // lines up with the edge of the original. That's OK. That just means that the subtraction didn't leave anything behind.
+                // We will filter those out below when adding them to the result.
+                const auto t = rectangle({ left(), top(), right(), intersect.top() });
+                const auto b = rectangle({ left(), intersect.bottom(), right(), bottom() });
+                const auto l = rectangle({ left(), intersect.top(), intersect.left(), intersect.bottom() });
+                const auto r = rectangle({ intersect.right(), intersect.top(), right(), intersect.bottom() });
+
+                if (!t.empty())
+                {
+                    result.push_back(t);
+                }
+
+                if (!b.empty())
+                {
+                    result.push_back(b);
+                }
+
+                if (!l.empty())
+                {
+                    result.push_back(l);
+                }
+
+                if (!r.empty())
+                {
+                    result.push_back(r);
+                }
+            }
+
+            return result;
+        }
+
+        constexpr ptrdiff_t top() const noexcept
+        {
+            return _topLeft.y();
+        }
+
+        template<typename T>
+        T top() const
+        {
+            T ret;
+            THROW_HR_IF(E_ABORT, !base::MakeCheckedNum(top()).AssignIfValid(&ret));
+            return ret;
+        }
+
+        constexpr ptrdiff_t bottom() const noexcept
+        {
+            return _bottomRight.y();
+        }
+
+        template<typename T>
+        T bottom() const
+        {
+            T ret;
+            THROW_HR_IF(E_ABORT, !base::MakeCheckedNum(bottom()).AssignIfValid(&ret));
+            return ret;
+        }
+
+        constexpr ptrdiff_t left() const noexcept
+        {
+            return _topLeft.x();
+        }
+
+        template<typename T>
+        T left() const
+        {
+            T ret;
+            THROW_HR_IF(E_ABORT, !base::MakeCheckedNum(left()).AssignIfValid(&ret));
+            return ret;
+        }
+
+        constexpr ptrdiff_t right() const noexcept
+        {
+            return _bottomRight.x();
+        }
+
+        template<typename T>
+        T right() const
+        {
+            T ret;
+            THROW_HR_IF(E_ABORT, !base::MakeCheckedNum(right()).AssignIfValid(&ret));
+            return ret;
+        }
+
+        ptrdiff_t width() const
+        {
+            ptrdiff_t ret;
+            THROW_HR_IF(E_ABORT, !(::base::MakeCheckedNum(right()) - left()).AssignIfValid(&ret));
+            return ret;
+        }

```

Comment:
ah because it could underflow/wrap?

Reply:
yep

Reply:
Nit: Isn't this what CheckSub is for?
`THROW_HR_IF(E_ABORT, !base::CheckSub(right(), left()).AssignIfValid(&ret));`

Reply:
Yeah, I think you're right, I was just... mushing this around until I settled on something and probably didn't choose the most optimal way of getting to the checked math. I think under the covers it will boil into the same thing.